SR-8 Add question endpoints

Mario-659 opened the pull request.

Mendrejk commented:
First, quick observations:
- Why is the RestController called QuestionResource and not QuestionController?
- You should not create the TeamRepository with this pull request. I know that you need it and we've stumbled into this issue with @juleczkatruskaweczka as well. There is a task on Jira to add the TeamRepository. It needs to be completed and merged before this (any endpoint related) PR.

I'll try to do a more in-depth review later today ;).

Mario-659 commented:
It's the convention at my company but I will rename it. 

Check out my tests. They were quite challenging 😅

Mendrejk commented on backend/recruitment-system/src/test/groovy/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/QuestionServiceSpec.groovy:
```diff
@@ -0,0 +1,208 @@
+    }
+
+    def "should return all questions from repository"() {
+        when:
```
add stories to all when and then blocks

Mario-659 replied:
Aren't tests already self explanatory? I don't really know what to write there. Look at this example
```groovy
    def "should return all questions from repository"() {
        when:
        def actual = questionService.getAllQuestions()

        then:
        1 * questionRepository.findAll() >> questions
        actual == questions
    }
```
I can already read this in plain English (*when questionService getAllQuestions method is called...*), but when I add story:
```groovy
    def "should return all questions from repository"() {
        when: "questionService getAllQuestions method is called"
        def actual = questionService.getAllQuestions()

        then:
        1 * questionRepository.findAll() >> questions
        actual == questions
    }
```
then when I read it it becomes more like this: *when questionService getAllQuestions method is called, questionService getAllQuestions method is called*
The same with then block

Mendrejk replied:
What about something like 
when "All questions are retrieved"
then "The QuestionRepository should be called once and the result should match with the FindAll result"?
But I guess you could argue that it's already self explanatory. If you don't want to then don't write the stories. I just always add stories and never though about whether it's already readable or not.

Mendrejk reviewed (changes requested):
Fix It!

Mario-659 commented on backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/QuestionService.kt:
```diff
@@ -0,0 +1,94 @@
+        val question = questionRepository.findById(questionId)
+        if (question.isPresent && question.get().team.id == teamId) {
+            questionRepository.delete(question.get())
+        }
```
Updated. Let me now if oneliners are fine or returns with curlies would be better

Mario-659 commented on backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/QuestionService.kt:
```diff
@@ -0,0 +1,94 @@
+        validateTeamId(teamId)
+
+        val questionToSave = Question(
+            0,
```
What do you mean? Can I omit id with default constructor? I added default values in constructors because something wasn't working properly (maybe serializing, I'm not sure now but I will check this later)

Mario-659 commented on backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/serializer/QuestionJsonSerializer.kt:
```diff
@@ -0,0 +1,28 @@
+package pl.edu.pwr.samorzad.informatyzacja.recruitmentsystem.serializer
+
+import com.fasterxml.jackson.core.JsonGenerator
+import com.fasterxml.jackson.core.JsonProcessingException
+import com.fasterxml.jackson.databind.JsonSerializer
+import com.fasterxml.jackson.databind.SerializerProvider
+import pl.edu.pwr.samorzad.informatyzacja.recruitmentsystem.model.Question
+import java.io.IOException
+
+class QuestionJsonSerializer : JsonSerializer<Question>() {
+    @Throws(IOException::class, JsonProcessingException::class)
+    override fun serialize(
+        question: Question, jsonGenerator: JsonGenerator,
+        serializerProvider: SerializerProvider?
+    ) {
+        with(jsonGenerator) {
+            writeStartObject()
+            writeNumberField("id", question.id)
+            writeStringField("content", question.content)
+            question.minimumMinutes
+                ?. let {writeNumberField("minimumMinutes", it) }
+                ?: writeNullField("minimumMinutes")
+            writeNumberField("maximumMinutes", question.maximumMinutes)
+            writeNumberField("teamId", question.team.id)
+            writeEndObject()
+        }
+    }
+}
```
Yeah, I tried to use adnotation but couldn't find the right one. This solution adds quite a lot of boilerplate code but on the other hand separates declaration of Entities from logic related to serializing

Mario-659 commented on backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/controller/QuestionResource.kt:
```diff
@@ -0,0 +1,41 @@
+import org.springframework.http.ResponseEntity
+import org.springframework.web.bind.annotation.*
+import pl.edu.pwr.samorzad.informatyzacja.recruitmentsystem.model.Question
+import pl.edu.pwr.samorzad.informatyzacja.recruitmentsystem.service.QuestionService
```
I will do that. Do we allow wildcard imports?

Mendrejk replied:
Sure, we can use wildcards. Maybe not for recruitmentsystem here as it's just 2 imports but it makes sense for springframework. Actually IntelliJ automatically wildcards imports, but I think 6 or more is the sweetspot when it starts wildcarding. I think it's better to just say import org.springframework.* rather than have a wall of fear everyone will just scroll over and never read

Bravo555 commented on backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/controller/QuestionController.kt:
```diff
@@ -0,0 +1,43 @@
+import org.slf4j.LoggerFactory
+
+@RestController
+@RequestMapping("/api/v1")
```
It's fine for now, but we should probably add a task for proper API version routing so it's not repeated for every endpoint

Bravo555 commented on backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/exception/TeamIdDoesNotMatchQuestionException.kt:
```diff
@@ -0,0 +1,6 @@
+import org.springframework.http.HttpStatus
+import org.springframework.web.server.ResponseStatusException
+
+class TeamIdDoesNotMatchQuestionException : ResponseStatusException(HttpStatus.CONFLICT, "Team id does not match the question id")
```
I think it makes more sense to return `404 NOT FOUND` here. This exception is used in `getQuestion` method, for which the endpoint is `teams/{teamId}/questions/{questionId}`. For a person using the endpoint, they don't care and they shouldn't be allowed to know whether the question exists in another team, and more importantly, it can be used to leak how many questions are in the database.

Mendrejk replied:
+1

Mendrejk commented on backend/recruitment-system/src/main/kotlin/pl/edu/pwr/samorzad/informatyzacja/recruitmentsystem/service/QuestionService.kt:
```diff
@@ -0,0 +1,103 @@
+        return questionRepository.findAllByTeamId(teamId).toList()
+    }
+
+    fun saveQuestion(question: Question, teamId: Int): Question {
+        validateQuestion(question)
+        validateTeamId(teamId)
+
+        val questionToSave = Question(
+            0,
+            question.content,
+            question.difficulty,
+            question.minimumMinutes,
+            question.maximumMinutes,
+            Team(teamId)
+        )
+
+        return questionRepository.save(questionToSave)
+    }
```
use addQuestion instead (also addQuesiton in repository)

Mendrejk reviewed (changes requested):
Fix It!

Mario-659 commented:
- added function in repository to simplify logic in service (@Mendrejk to return Nullable value you can override methods and specify return value instead of using `...OrNull` method but keep in mind that for collections it will return empty collection even if given id does not exist (from my own testing, I couldn't find that in docs 😞 ))
- @Bravo555 changed validation of question and team id, now it looks like this: assertTeamExists -> assertQuestionWithGivenTeamIdAndQuestionExists -> (for PUT and POST) validateQuestionBody
- updated tests but they still don't cover every possibility
- exceptions and controller implemented as in tutorial ([repo](https://github.com/petersommerhoff/thenewboston-spring-boot-with-kotlin/blob/main/src/main/kotlin/tv/codealong/tutorials/springboot/thenewboston/controller/BankController.kt))

To discuss:
1. Now when exception occurs in response body is only message from exception. I think much better would be something like default spring answer:
```
{
 "timestamp": 1500597044204,
 "status": 400,
 "error": "Bad Request",
 "exception": "org.springframework.http.converter.HttpMessageNotReadableException",
 "message": "message from exception",
 "path": "/birds"
}
```
I'm not sure how to do this but I can look into that or we can leave it as it is for now.
2. I validate RequestBody in service (`private fun validateQuestion(question: Question)`) however we can use something like [here](https://springframework.guru/bean-validation-in-spring-boot/)

Mendrejk reviewed (approved):
Ship It!

Mendrejk commented:
Regarding your suggestions, let's leave this as-is for now. We can create Jira's for them

Mendrejk commented:
> * added function in repository to simplify logic in service (@Mendrejk to return Nullable value you can override methods and specify return value instead of using `...OrNull` method but keep in mind that for collections it will return empty collection even if given id does not exist (from my own testing, I couldn't find that in docs disappointed ))
> * @Bravo555 changed validation of question and team id, now it looks like this: assertTeamExists -> assertQuestionWithGivenTeamIdAndQuestionExists -> (for PUT and POST) validateQuestionBody
> * updated tests but they still don't cover every possibility
> * exceptions and controller implemented as in tutorial ([repo](https://github.com/petersommerhoff/thenewboston-spring-boot-with-kotlin/blob/main/src/main/kotlin/tv/codealong/tutorials/springboot/thenewboston/controller/BankController.kt))
> 
> To discuss:
> 
> 1. Now when exception occurs in response body is only message from exception. I think much better would be something like default spring answer:
> 
> ```
> {
>  "timestamp": 1500597044204,
>  "status": 400,
>  "error": "Bad Request",
>  "exception": "org.springframework.http.converter.HttpMessageNotReadableException",
>  "message": "message from exception",
>  "path": "/birds"
> }
> ```
> 
> I'm not sure how to do this but I can look into that or we can leave it as it is for now. 2. I validate RequestBody in service (`private fun validateQuestion(question: Question)`) however we can use something like [here](https://springframework.guru/bean-validation-in-spring-boot/)

Jira [SR-27](https://informatyzacja.atlassian.net/browse/SR-27), [SR-28](https://informatyzacja.atlassian.net/browse/SR-28)
